src/service/userActivity.service.ts

```diff
@@ -35,16 +38,93 @@ export const createActivity = async ({
   }
 };
 
+// the tickets have three conditions:
+//   1. with _id
+//      - _id is existing tickets (update)
+//      - _id not found (throw new Error)
+//   2. ticket data without _id
+//      - create new tickets
+//   3. the existing tickets' _id are not pass to backend
+//      - delete the existing tickets if ticket is not sold
+//      - throw new Error if tickets is sold
+export const editActivity = async ({
+  userId,
+  activityId,
+  tickets,
+  ...activityData
+}: ActivityEditCredentials) => {
+  const existingActivity = await Activity.findById(activityId);
+  if (!existingActivity?.creatorId.equals(userId))
+    throw new CoreError('Activity cannot be edited by non-creators.');
+  const activityOrders = await Order.find({ activityId });
+  if (activityOrders.length) {
+    // todo：send notices to participant if tickets are sold
+  }
+
+  const existingTicketIds = (
+    await Ticket.find({ activityId }).select('_id')
+  ).map((ticket) => ticket._id.toString());
```

Comment:
這裡ticket._id就不要轉成string，以免你後面`updateTickets, createTickets, deleteTickets`id沒有存成ObjectId

Reply:
因為 objectId 無法使用 indexOf 來搜尋，update 是完全使用傳進來的資料做更新，create 有先將 activityId 轉成 objectId 再存，existingTicketIds 設計的目的是為了快速的找出本來活動的票券，但是沒有被傳到後端的，以用來在 delete 時做判斷一次刪除

Reply:
```
for (const deleteTicketId of existingTicketIds) {
      const deleteTicketOrders = await Order.find({
        ticketId: deleteTicketId,
      });
```
那這邊id是string，如果只要判斷index的話那就index這一行轉就好了？

Reply:
const ticketIndex = existingTicketIds.indexOf(ticket?._id.toString());
不太懂意思是這一行是指上面這一行嗎?
existingTicketIds 必須是 string array，才有辦法使用 indexOf 函數，然後會在80行
existingTicketIds.splice(ticketIndex, 1);
如果在更新名單內的話，就會排除在待刪除名單之外

就是 existingTicketIds 存 id string 陣列，是為了方便直接搜尋&判斷刪除
目的是為了找出前端傳進來的 tickets，是否有原本存在但沒傳進來的票券
existingTicketIds 會使用到的部份是 update & delete
因為 update 儲存的 id 是由前端傳進來的，所以不會用到這裡的 string id
delete 刪除時可直接用 string id 來讓 mongoose 辦認

Reply:
我的意思是existingTicketIds因為你已經map成string了，這邊loop你拿到的id都是string，Order.find因為你存的ticketId應該會是ObjectId，這樣會找不到
```
for (const deleteTicketId of existingTicketIds) {
      const deleteTicketOrders = await Order.find({
        ticketId: deleteTicketId,
      });
```

```suggestion
 const existingTicketObjectIds = (
    await Ticket.find({ activityId }).select('_id')
  )
  const ticketIndex = existingTicketObjectIds.map((id) => id.toString()).indexOf(ticket?._id.toString());
```

Reply:
哦哦 因為原本還沒有 Order，所以的確是沒測試到這個部份，已修正上傳